File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/GestureConsumer.java
package com.tunjid.fingergestures.gestureconsumers;

import androidx.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public interface GestureConsumer {

    int ZERO_PERCENT = 0;
    int FIFTY_PERCENT = 50;
    int HUNDRED_PERCENT = 100;

    int INCREASE_BRIGHTNESS = 0;
    int REDUCE_BRIGHTNESS = 1;
    int MAXIMIZE_BRIGHTNESS = 2;
    int MINIMIZE_BRIGHTNESS = 3;
    // int NIGHT_MODE_ON = 4; DO NOT REMOVE FOR LEGACY REASONS
    // int NIGHT_MODE_OFF = 5; DO NOT REMOVE FOR LEGACY REASONS
    int NOTIFICATION_UP = 6;
    int NOTIFICATION_DOWN = 7;
    int DO_NOTHING = 8;
    int TOGGLE_FLASHLIGHT = 9;
    int TOGGLE_DOCK = 10;
    int TOGGLE_AUTO_ROTATE = 11;
    int NOTIFICATION_TOGGLE = 12;
    int GLOBAL_HOME = 13;
    int GLOBAL_BACK = 14;
    int GLOBAL_RECENTS = 15;
    int GLOBAL_SPLIT_SCREEN = 16;
    int GLOBAL_POWER_DIALOG = 17;
    int SHOW_POPUP = 18;
    int INCREASE_AUDIO = 19;
    int REDUCE_AUDIO = 20;
    int GLOBAL_LOCK_SCREEN = 21;
    int GLOBAL_TAKE_SCREENSHOT = 22;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({INCREASE_BRIGHTNESS, REDUCE_BRIGHTNESS, MAXIMIZE_BRIGHTNESS, MINIMIZE_BRIGHTNESS,
            NOTIFICATION_DOWN, NOTIFICATION_UP, NOTIFICATION_TOGGLE, DO_NOTHING,
            TOGGLE_FLASHLIGHT, TOGGLE_DOCK, TOGGLE_AUTO_ROTATE,
            GLOBAL_HOME, GLOBAL_BACK, GLOBAL_RECENTS, GLOBAL_SPLIT_SCREEN, GLOBAL_POWER_DIALOG,
            SHOW_POPUP, REDUCE_AUDIO, INCREASE_AUDIO, GLOBAL_LOCK_SCREEN, GLOBAL_TAKE_SCREENSHOT})
    @interface GestureAction {}

    void onGestureActionTriggered(@GestureAction int gestureAction);

    boolean accepts(@GestureAction int gesture);

    static float normalizePercentageToFraction(int percentage) {
        return percentage / 100F;
    }

}

File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/GestureConsumer.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.gestureconsumers

import com.tunjid.fingergestures.R

/**
 * Actions available. The ids are serialized to shared preferences and MUST not change
 */
enum class GestureAction(val id: Int) {
    IncreaseBrightness(0),
    ReduceBrightness(1),
    MaximizeBrightness(2),
    MinimizeBrightness(3),
    // int NIGHT_MODE_ON = 4; DO NOT REMOVE FOR LEGACY REASONS
    // int NIGHT_MODE_OFF = 5; DO NOT REMOVE FOR LEGACY REASONS
    NotificationUp(6),
    NotificationDown(7),
    DoNothing(8),
    FlashlightToggle(9),
    DockToggle(10),
    AutoRotateToggle(11),
    NotificationToggle(12),
    GlobalHome(13),
    GlobalBack(14),
    GlobalRecents(15),
    GlobalSplitScreen(16),
    GlobalPowerDialog(17),
    PopUpShow(18),
    IncreaseAudio(19),
    ReduceAudio(20),
    GlobalLockScreen(21),
    GlobalTakeScreenshot(22);

    val serialized get() = id.toString()

    companion object {
        private val map = values().map { it.id to it }.toMap()
        fun fromId(id: Int) = map.getValue(id)
        fun deserialize(id: String) = fromId(id.toInt())
    }
}

val GestureAction.resource: Int
    get() = when (this) {
        GestureAction.DoNothing -> R.string.do_nothing
        GestureAction.IncreaseBrightness -> R.string.increase_brightness
        GestureAction.ReduceBrightness -> R.string.reduce_brightness
        GestureAction.MaximizeBrightness -> R.string.maximize_brightness
        GestureAction.MinimizeBrightness -> R.string.minimize_brightness
        GestureAction.IncreaseAudio -> R.string.increase_audio
        GestureAction.ReduceAudio -> R.string.reduce_audio
        GestureAction.NotificationUp -> R.string.notification_up
        GestureAction.NotificationDown -> R.string.notification_down
        GestureAction.NotificationToggle -> R.string.toggle_notifications
        GestureAction.FlashlightToggle -> R.string.toggle_flashlight
        GestureAction.DockToggle -> R.string.toggle_dock
        GestureAction.AutoRotateToggle -> R.string.toggle_auto_rotate
        GestureAction.GlobalHome -> R.string.global_home
        GestureAction.GlobalBack -> R.string.global_back
        GestureAction.GlobalRecents -> R.string.global_recents
        GestureAction.GlobalSplitScreen -> R.string.global_split_screen
        GestureAction.GlobalPowerDialog -> R.string.global_power_dialog
        GestureAction.GlobalLockScreen -> R.string.global_lock_screen
        GestureAction.GlobalTakeScreenshot -> R.string.global_take_screenshot
        GestureAction.PopUpShow -> R.string.show_popup
    }

interface GestureConsumer {

    fun onGestureActionTriggered(gestureAction: GestureAction)

    fun accepts(gesture: GestureAction): Boolean

    companion object {

        const val ZERO_PERCENT = 0
        const val FIFTY_PERCENT = 50
        const val HUNDRED_PERCENT = 100

        fun normalizePercentageToFraction(percentage: Int): Float = percentage / 100f
    }
}
